proxy/srs.go

```diff
@@ -0,0 +1,379 @@
+			servers = append(servers, server)
+		}
+		return true
+	})
+
+	// If no servers available, use all possible servers.
+	if len(servers) == 0 {
+		v.servers.Range(func(key string, server *SRSServer) bool {
+			servers = append(servers, server)
+			return true
+		})
+	}
+
+	// No server found, failed.
+	if len(servers) == 0 {
+		return nil, fmt.Errorf("no server available for %v", streamURL)
+	}
+
+	// Pick a server randomly from servers.
+	server := servers[rand.Intn(len(servers))]
+	v.picked.Store(streamURL, server)
+	return server, nil
+}
+
+type srsRedisLoadBalancer struct {
+	// The redis client sdk.
+	rdb *redis.Client
+}
+
+func NewRedisLoadBalancer() SRSLoadBalancer {
+	return &srsRedisLoadBalancer{}
+}
+
+func (v *srsRedisLoadBalancer) Initialize(ctx context.Context) error {
+	redisDatabase, err := strconv.Atoi(envRedisDB())
+	if err != nil {
+		return errors.Wrapf(err, "invalid PROXY_REDIS_DB %v", envRedisDB())
+	}
+
+	rdb := redis.NewClient(&redis.Options{
+		Addr:     fmt.Sprintf("%v:%v", envRedisHost(), envRedisPort()),
+		Password: envRedisPassword(),
+		DB:       redisDatabase,
+	})
+	v.rdb = rdb
+
+	if err := rdb.Ping(ctx).Err(); err != nil {
+		return errors.Wrapf(err, "unable to connect to redis %v", rdb.String())
+	}
+	logger.Df(ctx, "RedisLB: connected to redis %v ok", rdb.String())
+
+	if server, err := NewDefaultSRSForDebugging(); err != nil {
+		return errors.Wrapf(err, "initialize default SRS")
+	} else if server != nil {
+		if err := v.Update(ctx, server); err != nil {
+			return errors.Wrapf(err, "update default SRS %+v", server)
+		}
+
+		// Keep alive.
+		go func() {
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-time.After(30 * time.Second):
+					if err := v.Update(ctx, server); err != nil {
+						logger.Wf(ctx, "update default SRS %+v failed, %+v", server, err)
+					}
+				}
+			}
+		}()
+		logger.Df(ctx, "RedisLB: Initialize default SRS media server, %+v", server)
+	}
+	return nil
+}
+
+func (v *srsRedisLoadBalancer) Update(ctx context.Context, server *SRSServer) error {
+	b, err := json.Marshal(server)
+	if err != nil {
+		return errors.Wrapf(err, "marshal server %+v", server)
+	}
+
+	key := fmt.Sprintf("srs-proxy-server:%v", server.ID())
+	if err = v.rdb.Set(ctx, key, b, srsServerAliveDuration).Err(); err != nil {
+		return errors.Wrapf(err, "set key=%v server %+v", key, server)
+	}
+
+	// Query all servers from redis, in json string.
+	var serverKeys []string
+	if b, err := v.rdb.Get(ctx, v.redisKeyServers()).Bytes(); err == nil {
+		if err := json.Unmarshal(b, &serverKeys); err != nil {
+			return errors.Wrapf(err, "unmarshal key=%v servers %v", v.redisKeyServers(), string(b))
+		}
+	}
+
+	// Check each server expiration, if not exists in redis, remove from servers.
+	for i, serverKey := range serverKeys {
+		if _, err := v.rdb.Get(ctx, serverKey).Bytes(); err != nil {
+			serverKeys = append(serverKeys[:i], serverKeys[i+1:]...)
```

Comment:
remove a slice's element while iterating them by `range` is an error.
ref:
https://vbauerster.github.io/2017/04/removing-items-from-a-slice-while-iterating-in-go/
use:
https://pkg.go.dev/slices#DeleteFunc

Reply:
Fixed. Thanks!